### PR TITLE
upgrade sdk go from v1 to v2

### PR DESCRIPTION
**Pull request by outscale-hmi**
Signed-off-by: outscale_hmi <hanen.mizouni@outscale.com>

### CODE REVIEW

File: builder/osc/bsu/step_create_omi.go [hunk 1]
```diff
@@ -5,98 +5,91 @@ import (
...
+	ui.Message(fmt.Sprintf("OMI: %s", image.GetImageId()))
 	omis := make(map[string]string)
-	omis[s.RawRegion] = image.ImageId
+	omis[s.RawRegion] = image.GetImageId()
 	state.Put("omis", omis)
 
 	// Wait for the image to become ready
 	ui.Say("Waiting for OMI to become ready...")
-	if err := osccommon.WaitUntilOscImageAvailable(oscconn, image.ImageId); err != nil {
+	if err := osccommon.WaitUntilOscImageAvailable(oscconn, *image.ImageId); err != nil {
```
Comment (outscale-mdr) [inline builder/osc/bsu/step_create_omi.go:57]: use directly `image.getImageId()`

File: builder/osc/bsu/step_create_omi.go [hunk 2]
```diff
@@ -5,98 +5,91 @@ import (
...
+		}
+		imagesResp, _, err := oscconn.Api.ImageApi.ReadImages(oscconn.Auth).ReadImagesRequest(req).Execute()
 		if err != nil {
 			log.Printf("Unable to determine reason waiting for OMI failed: %s", err)
 			err = fmt.Errorf("Unknown error waiting for OMI")
 		} else {
-			stateReason := imagesResp.Images[0].StateComment
-			err = fmt.Errorf("Error waiting for OMI. Reason: %s", stateReason)
+			stateReason := imagesResp.GetImages()[0].GetStateComment()
+			err = fmt.Errorf("Error waiting for OMI. Reason: %s", *stateReason.StateMessage)
```
Comment (outscale-mdr) [inline builder/osc/bsu/step_create_omi.go:68]: Same here, `stateReason.GetStateMessage()`

File: builder/osc/bsu/step_create_omi.go [hunk 3]
```diff
@@ -5,98 +5,91 @@ import (
...
+	}
+	imagesResp, _, err := oscconn.Api.ImageApi.ReadImages(oscconn.Auth).ReadImagesRequest(req).Execute()
 	if err != nil {
 		err := fmt.Errorf("Error searching for OMI: %s", err)
 		state.Put("error", err)
 		ui.Error(err.Error())
 		return multistep.ActionHalt
 	}
-	s.image = &imagesResp.Images[0]
+	s.image = &imagesResp.GetImages()[0]
```
Comment (outscale-mdr) [inline builder/osc/bsu/step_create_omi.go:86]: We could add a test if the array is empty (in case)

File: builder/osc/bsu/step_create_omi.go [hunk 4]
```diff
@@ -5,98 +5,91 @@ import (
...
 
 	snapshots := make(map[string][]string)
-	for _, blockDeviceMapping := range imagesResp.Images[0].BlockDeviceMappings {
-		if blockDeviceMapping.Bsu.SnapshotId != "" {
-			snapshots[s.RawRegion] = append(snapshots[s.RawRegion], blockDeviceMapping.Bsu.SnapshotId)
+	blockMapping := imagesResp.GetImages()[0].BlockDeviceMappings
```
Comment (outscale-mdr) [inline builder/osc/bsu/step_create_omi.go:89]: `GetBlockDeviceMappings()`, this is usefull because if the var is  null, the getter will return an empty array see [link](https://github.com/outscale/osc-sdk-go/blob/7a11a0906c025139fb2b6f25cbd8f38890121333/v2/model_read_images_response.go#L43))

File: builder/osc/bsu/step_create_omi.go [hunk 5]
```diff
@@ -115,14 +108,13 @@ func (s *stepCreateOMI) Cleanup(state multistep.StateBag) {
...
-	oscconn := state.Get("osc").(*osc.APIClient)
+	oscconn := state.Get("osc").(*osccommon.OscClient)
 	ui := state.Get("ui").(packersdk.Ui)
 
 	ui.Say("Deregistering the OMI because cancellation or error...")
-	DeleteOpts := osc.DeleteImageRequest{ImageId: s.image.ImageId}
-	if _, _, err := oscconn.ImageApi.DeleteImage(context.Background(), &osc.DeleteImageOpts{
-		DeleteImageRequest: optional.NewInterface(DeleteOpts),
-	}); err != nil {
+	DeleteOpts := oscgo.DeleteImageRequest{ImageId: s.image.GetImageId()}
```
Comment (outscale-mdr) [inline builder/osc/bsu/step_create_omi.go:115]: we can change to `deleteOpts`, it was the previous code that has that

File: builder/osc/bsusurrogate/step_register_omi.go [hunk 6]
```diff
@@ -4,86 +4,81 @@ import (
...
-
-	registerOpts := osc.CreateImageRequest{
-		ImageName:           config.OMIName,
-		Architecture:        "x86_64",
-		RootDeviceName:      s.RootDevice.DeviceName,
-		BlockDeviceMappings: blockDevices,
+	architecture := "x86_64"
+	registerOpts := oscgo.CreateImageRequest{
+		ImageName:           &config.OMIName,
+		Architecture:        &architecture,
```
Comment (outscale-mdr) [inline builder/osc/bsusurrogate/step_register_omi.go:35]: I've discovered that the SDK have a Utils package which contains that:
```func PtrString(v string) *string { return &v }```(see [link](https://github.com/outscale/osc-sdk-go/blob/7a11a0906c025139fb2b6f25cbd8f38890121333/v2/utils.go#L38)).

I do not ask for change, just to let you know

File: builder/osc/bsusurrogate/step_register_omi.go [hunk 7]
```diff
@@ -102,62 +97,59 @@ func (s *StepRegisterOMI) Cleanup(state multistep.StateBag) {
...
-		if device.DeviceName == s.RootDevice.SourceDeviceName {
-			device.DeviceName = s.RootDevice.DeviceName
+		if device.GetDeviceName() == s.RootDevice.SourceDeviceName {
+			device.DeviceName = &s.RootDevice.DeviceName
 
-			if device.Bsu.VolumeType != "" {
-				device.Bsu.VolumeType = s.RootDevice.VolumeType
-				if device.Bsu.VolumeType != "io1" {
-					device.Bsu.Iops = 0
+			if device.Bsu.VolumeType != nil {
```
Comment (outscale-mdr) [inline builder/osc/bsusurrogate/step_register_omi.go:130]: This could be changed to ```if _, ok device.Bsu.GetVolumeTypeOk(); ok {```

File: builder/osc/bsuvolume/step_tag_bsu_volumes.go [hunk 8]
```diff
@@ -53,22 +52,21 @@ func (s *stepTagBSUVolumes) Run(_ context.Context, state multistep.StateBag) mul
...
-				CreateTagsRequest: optional.NewInterface(osc.CreateTagsRequest{
-					ResourceIds: []string{volumeId},
-					Tags:        tags,
-				}),
-			})
+			request := oscgo.CreateTagsRequest{
+				ResourceIds: []string{volumeId},
+				Tags:        tags,
+			}
+			_, _, err := oscconn.TagApi.CreateTags(context.Background()).CreateTagsRequest(request).Execute()
```
Comment (outscale-mdr) [inline builder/osc/bsuvolume/step_tag_bsu_volumes.go:67]: Change the ```context.background()```, right ?

File: builder/osc/chroot/step_create_omi.go [hunk 9]
```diff
@@ -19,85 +18,85 @@ type StepCreateOMI struct {
...
 	if config.FromScratch {
 		mappings = config.OMIBlockDevices.BuildOscOMIDevices()
 		rootDeviceName = config.RootDeviceName
 	} else {
-		image = state.Get("source_image").(osc.Image)
-		mappings = image.BlockDeviceMappings
-		rootDeviceName = image.RootDeviceName
+		image = state.Get("source_image").(oscgo.Image)
+		mappings = *image.BlockDeviceMappings
+		rootDeviceName = *image.RootDeviceName
```
Comment (outscale-mdr) [inline builder/osc/chroot/step_create_omi.go:41]: Use the getter, it is easier

File: builder/osc/chroot/step_create_omi.go [hunk 10]
```diff
@@ -19,85 +18,85 @@ type StepCreateOMI struct {
...
-		gibSize := newDevice.Bsu.VolumeSize / (1024 * 1024 * 1024)
-		newDevice.Bsu.VolumeSize = gibSize
+		gibSize := *newDevice.Bsu.VolumeSize / (1024 * 1024 * 1024)
+		newDevice.Bsu.VolumeSize = &gibSize
 
-		if newDevice.DeviceName == rootDeviceName {
-			if newDevice.Bsu != (osc.BsuToCreate{}) {
-				newDevice.Bsu.SnapshotId = snapshotId
+		if newDevice.GetDeviceName() == rootDeviceName {
+			if *newDevice.Bsu != (oscgo.BsuToCreate{}) {
```
Comment (outscale-mdr) [inline builder/osc/chroot/step_create_omi.go:53]: If think, you just need to check that the newDevice.Bsu is nil or not. The v1 sdk, we were forced to check values, but now just check the pointer

File: builder/osc/chroot/step_create_omi.go [hunk 11]
```diff
@@ -19,85 +18,85 @@ type StepCreateOMI struct {
...
-	})
+	registerResp, _, err := osconn.Api.ImageApi.CreateImage(osconn.Auth).CreateImageRequest(registerOpts).Execute()
 	if err != nil {
 		state.Put("error", fmt.Errorf("Error registering OMI: %s", err))
 		ui.Error(state.Get("error").(error).Error())
 		return multistep.ActionHalt
 	}
 
-	imageID := registerResp.Image.ImageId
+	imageID := registerResp.GetImage().ImageId
```
Comment (outscale-mdr) [inline builder/osc/chroot/step_create_omi.go:90]: `GetImageId()`

File: builder/osc/chroot/step_create_volume.go [hunk 12]
```diff
@@ -127,47 +125,45 @@ func (s *StepCreateVolume) Cleanup(state multistep.StateBag) {
...
-func (s *StepCreateVolume) buildCreateVolumeInput(suregionName string, rootDevice *osc.BlockDeviceMappingImage) (*osc.CreateVolumeRequest, error) {
+func (s *StepCreateVolume) buildCreateVolumeInput(suregionName string, rootDevice *oscgo.BlockDeviceMappingImage) (*oscgo.CreateVolumeRequest, error) {
 	if rootDevice == nil {
 		return nil, fmt.Errorf("Couldn't find root device")
 	}
 
 	//FIX: Temporary fix
-	gibSize := rootDevice.Bsu.VolumeSize / (1024 * 1024 * 1024)
-	createVolumeInput := &osc.CreateVolumeRequest{
+	gibSize := *rootDevice.Bsu.VolumeSize / (1024 * 1024 * 1024)
```
Comment (outscale-mdr) [inline builder/osc/chroot/step_create_volume.go:144]: use the getter

File: builder/osc/chroot/step_snapshot.go [hunk 13]
```diff
@@ -44,7 +42,7 @@ func (s *StepSnapshot) Run(ctx context.Context, state multistep.StateBag) multis
 	}
 
 	// Set the snapshot ID so we can delete it later
-	s.snapshotId = createSnapResp.Snapshot.SnapshotId
+	s.snapshotId = *createSnapResp.GetSnapshot().SnapshotId
```
Comment (outscale-mdr) [inline builder/osc/chroot/step_snapshot.go:45]: use the getter

File: builder/osc/common/access_config.go [hunk 14]
```diff
@@ -124,13 +130,21 @@ func (c *AccessConfig) NewOSCClientByRegion(region string) *osc.APIClient {
...
-		DefaultHeader: make(map[string]string),
-		UserAgent:     fmt.Sprintf("packer-osc/%s", version.PluginVersion.String()),
-		HTTPClient:    skipClient,
-		Debug:         true,
+	config := oscgo.NewConfiguration()
+	config.Debug = true
+	config.UserAgent = fmt.Sprintf("packer-plugin-outscale/%s", version.PluginVersion.String())
+	auth := context.WithValue(context.Background(), oscgo.ContextAWSv4, osc.AWSv4{
+		AccessKey: os.Getenv("OSC_ACCESS_KEY"),
+		SecretKey: os.Getenv("OSC_SECRET_KEY"),
```
Comment (outscale-mdr) [inline builder/osc/common/access_config.go:138]: The TLS part was forgotten, add ``` config.HTTPClient = skipClient```

File: builder/osc/common/block_device.go [hunk 15]
```diff
@@ -34,87 +35,89 @@ type LaunchBlockDevices struct {
...
 			}
 
 			if blockDevice.VolumeType != "" {
-				bsu.VolumeType = blockDevice.VolumeType
+				bsu.VolumeType = &blockDevice.VolumeType
 			}
 
 			if blockDevice.VolumeSize > 0 {
-				bsu.VolumeSize = int32(blockDevice.VolumeSize)
+				bsu.VolumeSize = aws.Int32(int32(blockDevice.VolumeSize))
```
Comment (outscale-mdr) [inline builder/osc/common/block_device.go:60]: Use [this](https://github.com/outscale/osc-sdk-go/blob/7a11a0906c025139fb2b6f25cbd8f38890121333/v2/utils.go#L26) instead of AWS

File: builder/osc/common/block_device.go [hunk 16]
```diff
@@ -34,87 +35,89 @@ type LaunchBlockDevices struct {
...
 			}
 
 			// IOPS is only valid for io1 type
 			if blockDevice.VolumeType == "io1" {
-				bsu.Iops = int32(blockDevice.IOPS)
+				bsu.Iops = aws.Int32(int32(blockDevice.IOPS))
```
Comment (outscale-mdr) [inline builder/osc/common/block_device.go:65]: same

File: builder/osc/common/block_device.go [hunk 17]
```diff
@@ -34,87 +35,89 @@ type LaunchBlockDevices struct {
...
 			}
 
 			if blockDevice.VolumeType != "" {
-				bsu.VolumeType = blockDevice.VolumeType
+				bsu.VolumeType = &blockDevice.VolumeType
 			}
 
 			if blockDevice.VolumeSize > 0 {
-				bsu.VolumeSize = int32(blockDevice.VolumeSize)
+				bsu.VolumeSize = aws.Int32(int32(blockDevice.VolumeSize))
```
Comment (outscale-mdr) [inline builder/osc/common/block_device.go:107]: same

File: builder/osc/common/block_device.go [hunk 18]
```diff
@@ -34,87 +35,89 @@ type LaunchBlockDevices struct {
...
+
 			}
 
 			// IOPS is only valid for io1 type
 			if blockDevice.VolumeType == "io1" {
-				bsu.Iops = int32(blockDevice.IOPS)
+				bsu.Iops = aws.Int32(int32(blockDevice.IOPS))
```
Comment (outscale-mdr) [inline builder/osc/common/block_device.go:113]: same

File: builder/osc/common/block_device_test.go [hunk 19]
```diff
@@ -21,13 +23,13 @@ func TestBlockDevice_LaunchDevices(t *testing.T) {
...
-					VolumeType:         "standard",
-					VolumeSize:         8,
-					DeleteOnVmDeletion: true,
+			Result: oscgo.BlockDeviceMappingVmCreation{
+				DeviceName: aws.String("/dev/sdb"),
+				Bsu: &oscgo.BsuToCreate{
+					SnapshotId:         aws.String("snap-1234"),
+					VolumeType:         aws.String("standard"),
+					VolumeSize:         aws.Int32(8),
+					DeleteOnVmDeletion: aws.Bool(true),
```
Comment (outscale-mdr) [inline builder/osc/common/block_device_test.go:32]: use sdk utils fonction instead of aws

File: builder/osc/common/block_device_test.go [hunk 20]
```diff
@@ -37,11 +39,11 @@ func TestBlockDevice_LaunchDevices(t *testing.T) {
...
-			Result: osc.BlockDeviceMappingVmCreation{
-				DeviceName: "/dev/sdb",
-				Bsu: osc.BsuToCreate{
-					VolumeSize:         8,
-					DeleteOnVmDeletion: false,
+			Result: oscgo.BlockDeviceMappingVmCreation{
+				DeviceName: aws.String("/dev/sdb"),
+				Bsu: &oscgo.BsuToCreate{
+					VolumeSize:         aws.Int32(8),
+					DeleteOnVmDeletion: aws.Bool(false),
```
Comment (outscale-mdr) [inline builder/osc/common/block_device_test.go:46]: same

File: builder/osc/common/block_device_test.go [hunk 21]
```diff
@@ -54,13 +56,13 @@ func TestBlockDevice_LaunchDevices(t *testing.T) {
...
-					VolumeSize:         8,
-					DeleteOnVmDeletion: true,
-					Iops:               1000,
+			Result: oscgo.BlockDeviceMappingVmCreation{
+				DeviceName: aws.String("/dev/sdb"),
+				Bsu: &oscgo.BsuToCreate{
+					VolumeType:         aws.String("io1"),
+					VolumeSize:         aws.Int32(8),
+					DeleteOnVmDeletion: aws.Bool(true),
+					Iops:               aws.Int32(1000),
```
Comment (outscale-mdr) [inline builder/osc/common/block_device_test.go:65]: same

File: builder/osc/common/state.go [hunk 22]
```diff
@@ -100,168 +96,146 @@ func waitUntilOscVmStateFunc(conn *osc.APIClient, id string) stateRefreshFunc {
...
 		}
 
-		if resp.Images[0].State == "failed" {
-			return resp.Images[0].State, fmt.Errorf("Image (%s) creation is failed", id)
+		if *resp.GetImages()[0].State == "failed" {
+			return *resp.GetImages()[0].State, fmt.Errorf("Image (%s) creation is failed", id)
 		}
 
-		return resp.Images[0].State, nil
+		return *resp.GetImages()[0].State, nil
```
Comment (outscale-mdr) [inline builder/osc/common/state.go:193]: use getter

File: builder/osc/common/step_cleanup_volumes.go [hunk 23]
```diff
@@ -44,22 +43,17 @@ func (s *StepCleanupVolumes) Cleanup(state multistep.StateBag) {
 	// to device name, to compare with save list below
 	var vl []string
 	volList := make(map[string]string)
-	for _, bdm := range vm.BlockDeviceMappings {
-		if !reflect.DeepEqual(bdm.Bsu, osc.BsuCreated{}) {
-			vl = append(vl, bdm.Bsu.VolumeId)
-			volList[bdm.Bsu.VolumeId] = bdm.DeviceName
+	for _, bdm := range *vm.BlockDeviceMappings {
+		if !reflect.DeepEqual(bdm.Bsu, oscgo.BsuCreated{}) {
```
Comment (outscale-mdr) [inline builder/osc/common/step_cleanup_volumes.go:47]: Best to check the pointer, no ?

File: builder/osc/common/step_create_tags.go [hunk 24]
```diff
@@ -88,31 +84,29 @@ func (s *StepCreateTags) Run(_ context.Context, state multistep.StateBag) multis
...
+				Tags:        omiTags,
+			}
+
+			_, _, err := regionconn.Api.TagApi.CreateTags(regionconn.Auth).CreateTagsRequest(request).Execute()
 			if awsErr, ok := err.(awserr.Error); ok {
 				if awsErr.Code() == "InvalidOMIID.NotFound" ||
 					awsErr.Code() == "InvalidSnapshot.NotFound" {
 					return false, nil
 				}
 			}
```
Comment (outscale-mdr) [inline builder/osc/common/step_create_tags.go:101]: These will not work anymore, we could plan to update them

File: builder/osc/common/step_source_omi_info.go [hunk 25]
```diff
@@ -75,43 +74,44 @@ func (s *StepSourceOMIInfo) Run(_ context.Context, state multistep.StateBag) mul
 				oali = append(oali, o)
 			}
 		}
-		params.Filters.AccountIds = oid
-		params.Filters.AccountAliases = oali
+		//params.Filters.SetAccountIds(oid)
+		//params.Filters.SetAccountAliases(oali)
```
Comment (outscale-mdr) [inline builder/osc/common/step_source_omi_info.go:78]: remove comments

**Review by outscale-mdr — CHANGES_REQUESTED** [review]
Good job

File: builder/osc/bsu/step_create_omi.go (hunk 3 above)
Comment (outscale-hmi) [reply builder/osc/bsu/step_create_omi.go:86]: done

File: builder/osc/common/step_create_tags.go (hunk 24 above)
Comment (outscale-hmi) [reply builder/osc/common/step_create_tags.go:101]: I ll do it in a seperate PR

Comment (outscale-mdr) [reply builder/osc/common/step_create_tags.go:101]: Ok good

**Review by outscale-mdr — CHANGES_REQUESTED** [review]
Good job, just add this check and we will merge it 👍

**Review by outscale-mdr — APPROVED** [review]
